src/wp-includes/block-patterns.php

```diff
@@ -319,37 +319,132 @@ function _register_remote_theme_patterns() {
+				continue;
+			}
+
+			// The actual pattern content is the output of the file.
+			ob_start();
+			include $dirpath . $file;
+			$pattern_data['content'] = ob_get_clean();
+			if ( ! $pattern_data['content'] ) {
+				continue;
+			}
+
+			// Translate the pattern metadata.
+			// phpcs:ignore WordPress.WP.I18n.NonSingularStringLiteralText,WordPress.WP.I18n.NonSingularStringLiteralDomain,WordPress.WP.I18n.LowLevelTranslationFunction
+			$pattern_data['title'] = translate_with_gettext_context( $pattern_data['title'], 'Pattern title', $text_domain );
+			if ( ! empty( $pattern_data['description'] ) ) {
+				// phpcs:ignore WordPress.WP.I18n.NonSingularStringLiteralText,WordPress.WP.I18n.NonSingularStringLiteralDomain,WordPress.WP.I18n.LowLevelTranslationFunction
+				$pattern_data['description'] = translate_with_gettext_context( $pattern_data['description'], 'Pattern description', $text_domain );
+			}
+
+			register_block_pattern( $pattern_data['slug'], $pattern_data );
+		}
+	}
+}
+add_action( 'init', '_register_theme_block_patterns' );
+
+/**
+ * Gets block pattern data for a specified theme.
+ * Each pattern is defined as a PHP file and defines
+ *  its metadata using plugin-style headers. The minimum required definition is:
  *
- * The output of the PHP source corresponds to the content of the pattern, e.g.:
+ *      /**
+ *       * Title: My Pattern
+ *       * Slug: my-theme/my-pattern
+ *       *
  *
- *     <main><p><?php echo "Hello"; ?></p></main>
+ *  The output of the PHP source corresponds to the content of the pattern, e.g.:
  *
- * If applicable, this will collect from both parent and child theme.
+ *      <main><p><?php echo "Hello"; ?></p></main>
  *
- * Other settable fields include:
+ *  If applicable, this will collect from both parent and child theme.
  *
- *   - Description
- *   - Viewport Width
- *   - Inserter         (yes/no)
- *   - Categories       (comma-separated values)
- *   - Keywords         (comma-separated values)
- *   - Block Types      (comma-separated values)
- *   - Post Types       (comma-separated values)
- *   - Template Types   (comma-separated values)
+ *  Other settable fields include:
  *
- * @since 6.0.0
- * @since 6.1.0 The `postTypes` property was added.
- * @since 6.2.0 The `templateTypes` property was added.
+ *    - Description
+ *    - Viewport Width
+ *    - Inserter         (yes/no)
+ *    - Categories       (comma-separated values)
+ *    - Keywords         (comma-separated values)
+ *    - Block Types      (comma-separated values)
+ *    - Post Types       (comma-separated values)
+ *    - Template Types   (comma-separated values)
+ *
+ * @since 6.4.0
  * @access private
+ *
+ * @param WP_Theme $theme Theme object.
+ * @return array Block pattern data.
  */
-function _register_theme_block_patterns() {
+
+function _wp_get_block_patterns( WP_Theme $theme ) {
+	if ( ! $theme->exists() ) {
+		return array(
+			'version'  => false,
+			'patterns' => array(),
+		);
+	}
```

Comment:
Returning the full `version` plus `patterns` array here is bad API design. The version is only here for internal usage to ensure the cached value is not stale. We should only return `$pattern_data['patterns']` throughout this function, otherwise this function is confusing.

See also my related comment for abstracting the cache specific logic into `WP_Theme` alongside the function you already created.